src/runtime/hashmap.go

```diff
@@ -57,12 +68,70 @@ func hashmapMake(keySize, valueSize uint8, sizeHint uintptr) *hashmap {
 	}
 	bucketBufSize := unsafe.Sizeof(hashmapBucket{}) + uintptr(keySize)*8 + uintptr(valueSize)*8
 	buckets := alloc(bucketBufSize*(1<<bucketBits), nil)
+
+	keyHash := hashmapKeyHashAlg(hashmapAlgorithm(alg))
+	keyEqual := hashmapKeyEqualAlg(hashmapAlgorithm(alg))
+
 	return &hashmap{
 		buckets:    buckets,
 		keySize:    keySize,
 		valueSize:  valueSize,
 		bucketBits: bucketBits,
+		keyEqual:   keyEqual,
+		keyHash:    keyHash,
+	}
+}
+
+func hashmapKeyEqualAlg(alg hashmapAlgorithm) func(x, y unsafe.Pointer, n uintptr) bool {
+	switch alg {
+	case hashmapAlgorithmBinary:
+		return memequal
+	case hashmapAlgorithmString:
+		return hashmapStringEqual
+	case hashmapAlgorithmInterface:
+		return hashmapInterfaceEqual
+	default:
+		panic("unknown hashmap equal algorithm")
+	}
+}
+
+func hashmapKeyHashAlg(alg hashmapAlgorithm) func(key unsafe.Pointer, n uintptr) uint32 {
+	switch alg {
+	case hashmapAlgorithmBinary:
+		return hash32
+	case hashmapAlgorithmString:
+		return hashmapStringPtrHash
+	case hashmapAlgorithmInterface:
+		return hashmapInterfacePtrHash
+	default:
+		panic("unknown hashmap hash algorithm")
+	}
+}
+
+func hashmapShouldGrow(m *hashmap) bool {
+
+	if heapStart == heapEnd {
+		// don't try to grow if we have no memory
+		// this also triggers during interp
+		return false
 	}
```

Comment:
I don't agree with this. I'm pretty sure this indicates a bug somewhere else in the code.
(I need to review the rest of the code still, it's probably fine)

Reply:
OK, so what do we do now? If this indicates a bug elsewhere, is there a way we can trigger it and find it?

If we are not going to do this right now, we could create an issue to track doing so, and then merge this PR (pending any other feedback) in the interest of unblocking this feature.

Reply:
I have a strong suspicion this is an interp bug. One possible solution (that's still a hack but a lot cleaner) is to add `runtime.hashmapGrow` to the functions that are excluded from interp:

https://github.com/tinygo-org/tinygo/blob/9eb4a6268aa8d61de9d6c03d6d02c0f0c2be157f/interp/interpreter.go#L204-L217

Reply:
```
heapStart: 0x0009fff0
heapEnd: 0x000a0000
```
Actually, it might be a GC issue after all. These two addresses are awkwardly close together.

Reply:
I think I have a fix. I'll test it a bit more to confirm.

Reply:
@dgryski #2812 is in `dev` now can you please rebase and make suitable adjustments to this PR?